src/gui/guiiconprovider.cpp

```diff
@@ -135,13 +148,87 @@ QString GuiIconProvider::getIconPath(const QString &iconId)
         return path;
     }
 #endif
-    return IconProvider::getIconPath(iconId);
+    return IconProvider::getIconPath(iconId, overrideTheme);
 }
 
-
 void GuiIconProvider::configure()
 {
 #if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC))
     m_useSystemTheme = Preferences::instance()->useSystemIconTheme();
 #endif
 }
+
+int GuiIconProvider::getThemeFlags()
+{
+    // QPalette::Base is used for the background of the Treeview
+    QPalette pal = QApplication::palette();
+    QColor color = pal.color(QPalette::Active, QPalette::Base);
+    bool darkTheme = (color.lightness() < 127);
+    int themeFlags = darkTheme ? DarkTheme : LightTheme;
+
+    const Preferences *const pref = Preferences::instance();
+    themeFlags |= pref->useLargeStatusIcons() ? LargeStatusIcons : SmallStatusIcons;
+
+    return themeFlags;
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId)
+{
+    return getStatusIcon(iconId, BitTorrent::TorrentState::Unknown, true);
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId, BitTorrent::TorrentState state)
+{
+    return getStatusIcon(iconId, state, false);
+}
+
+QIcon GuiIconProvider::getStatusIcon(const QString &iconId, BitTorrent::TorrentState state, bool ignoreState)
+{
+    static int cachedIconFlags = 0;
+    static QHash<QPair<QString, BitTorrent::TorrentState>, QIcon> cache;
+    int theme = getThemeFlags();
+
+    if (cachedIconFlags != theme) {
+        qDebug().nospace() << "Changing status icon theme to " << ((theme & DarkTheme) ? "dark" : "light")
+                           << ", " << ((theme & LargeStatusIcons) ? "large" : "small") << " icons";
+        cache.clear();
+        cachedIconFlags = theme;
+    }
+
+    QString path(getIconPath(iconId, "skin"));
+    QPair<QString, BitTorrent::TorrentState> key(path, state);
+    QIcon icon(cache[key]);
+
+    if (!icon.isNull())
+        return icon;
+
+    QPixmap pixmap(path);
+    QPixmap &iconPixmap(pixmap);
```

Comment:
> which may be reassigned to colored below, rather than unnecessarily copying objects. 

This is wrong in C++. One needs a pointer for this kind of things.